[DPE-6466] Create the benchmark docs (#310)
canonical/kafka-operator: docs/how-to/h-benchmark.md

# How to benchmark Charmed Apache Kafka

The Apache Kafka benchmark charm uses the OpenMessaging tool to test both producer and consumer performance in the cluster.

The OpenMessaging allows for a distributed deployment, where the charm leader will run the main "manager" process and gather the metrics, whilst other units will act as followers and act as producer/consumers of the cluster.

## Deploy

Apache Kafka benchmark charm can be connected with either the [Charmed Apacke Kafka](https://canonical.com/data/docs/kafka/iaas) or the [Charmed Apache Kafka K8s](https://canonical.com/data/docs/kafka/k8s).

### Apache Kafka setup

#### Baremetal or Virtual Machines

Deploy Charmed Apache Kafka following the instructions in its [deployment guide](https://canonical.com/data/docs/kafka/iaas/h-deploy)

#### Kubernetes

Deploy Charmed Apache Kafka following the instructions in its [deployment guide](https://canonical.com/data/docs/kafka/k8s/t-deploy).

The Apache Kafka cluster must be configured to expose its client endpoints via NodePort, [æs described here](https://canonical.com/data/docs/kafka/k8s/h-external-k8s-connection).

> ### Benchmark Deployment

Reviewer: ```suggestion
### Benchmark deployment
```
We use sentence case as per our [style guide](https://docs.ubuntu.com/styleguide/en#headings)

Add the Apache Kafka benchmark charm using:

```
juju deploy kafka-benchmark --channel=latest/edge
juju relate kafka kafka-benchmark
```

### Integrate with COS

The benchmark charm needs a [COS](https://charmhub.io/topics/canonical-observability-stack) deployment running on a separate Juju model in Kubernetes to collect metrics during the test. Relate the Apache Kafka benchmark with a [grafana-agent operator](https://charmhub.io/grafana-agent).

For instructions on how to deploy and configure COS and its agents, see the [How to enable monitoring](https://canonical.com/data/docs/kafka/iaas/h-enable-monitoring) guide.

Once the grafana-agent is deployed, relate it with:

```
juju relate grafana-agent kafka-benchmark
```

The benchmark data will be collected every 10s and sent to prometheus.

### Integrate with TLS

Optionally, the entire deployment can use TLS.

> Charmed Apache Kafka supports TLS integration and described for both [baremetal / virtual machines](https://canonical.com/data/docs/kafka/k8s/t-enable-encryption) or [k8s](https://canonical.com/data/docs/kafka/iaas/h-enable-encryption) operators.

Reviewer: See the next comment. Additionally to the previous suggestion, to remove the next paragraph:
```suggestion
Charmed Apache Kafka supports TLS provider (such as [self-signed-certificates](https://charmhub.io/self-signed-certificates)) integration as described for both [machine](https://canonical.com/data/docs/kafka/iaas/h-enable-encryption) and [k8s](https://canonical.com/data/docs/kafka/k8s/h-enable-encryption) charmed operators, for example:
```

The Apache Kafka Benchmark can be then related to a TLS operator, such as [self-signed-certificates](https://charmhub.io/self-signed-certificates), for example:

```
juju integrate kafka-benchmark self-signed-certificates
```

> ## Start the Benchmark

Reviewer: **Suggestion**: We do have stop and clean-up in this section so it's not only about starting but more like the whole process of execution. The "use" verb doesn't sit right here with me, so `run` seems good enough even with stop inside (it's all part of its lifecycle). Let's aim for the following structure of the page:

* Deploy
* Run the benchmark
* Get the result

P.S. We are using the sentence case for headings (as per our [style guide](https://docs.ubuntu.com/styleguide/en#headings)).

```suggestion
## Run the benchmark
```

### Prepare

> The first step is to prepare the benchmark fleet with:

Reviewer: **Question**: Can we add some more details on this? Can we configure the number of units to be used or something like that? Can be very useful to add some details on how to configure the benchmark parameters.

Author: We can, but then that is at "deployment" time right. I will add that info there.

```
juju run kafka-benchmark/leader prepare
```

> ### Run

Reviewer: **Suggestion**: Let's use `Start` here to avoid repeating the newly suggested parent heading and to be more consistent with the next `stop` heading.

```suggestion
### Start
```

> The next command is `run`, to start the benchmark charms to produce and consume data.

Reviewer: **Suggestion**:
```suggestion
Use the `run` action, to start benchmarking (to produce and consume data):
```

```
juju run kafka-benchmark/leader run
```

The units will pick-up the command and start executing the benchmark.

### Stop

To stop the benchmark, execute:

```
juju run kafka-benchmark/leader stop
```

Optionally, it is possible to clean the current benchmark data using:

### Clean-up

To return the cluster to its original state, run:

```
juju run kafka-benchmark/leader cleanup
```

> ## Get the Results

Reviewer: ```suggestion
## Get the results
```

Kafka benchmark charm updates its metrics to COS and can be found under `openmessaging_` metric on prometheus:

> ![Screenshot from 2025-02-14 11-02-47](https://github.com/user-attachments/assets/b9da658c-1d76-4f2b-9cbc-f4243123cb34)

Reviewer: **Suggestion**: Improve image's name
```suggestion
![Prometheus results example](https://github.com/user-attachments/assets/b9da658c-1d76-4f2b-9cbc-f4243123cb34)
```

Author: I did not noticed when I copied :D thanks! That was a good find hehe

# How to benchmark Charmed Apache Kafka

> The Apache Kafka benchmark charm uses the OpenMessaging tool to test both producer and consumer performance in the cluster.

Reviewer: Suggestion: Let's add the link to the benchmark charm somewhere. Otherwise it's rather hard to find, since the charm is not searchable on the charmhub.
```suggestion
[The Apache Kafka benchmark](https://charmhub.io/kafka-benchmark) charm uses the OpenMessaging tool to test both producer and consumer performance in the cluster.
```

> # How to benchmark Charmed Apache Kafka

Reviewer: ```suggestion
# How to benchmark Charmed Apache Kafka

```
**Nitpick**: Please leave an empty line between each heading and the rest of the content.

The Apache Kafka benchmark charm uses the OpenMessaging tool to test both producer and consumer performance in the cluster.

The OpenMessaging allows for a distributed deployment, where the charm leader will run the main "manager" process and gather the metrics, whilst other units will act as followers and act as producer/consumers of the cluster.

## Deploy

> Create a new Juju model:

Reviewer: Since we don't have anything specific in this setup, we can start from a "fresh" Kafka deployment as a pre-requisite with a reference to the Kafka's [How to deploy](https://canonical.com/data/docs/kafka/iaas/h-deploy) guide.

```
juju add-model kafka-benchmark

> Deploy Apache Kafka with Juju:

Reviewer: This doesn't look like a command.
**todo**: fix the code blocks to contain only commands.

juju deploy kafka --channel=3/edge
juju deploy zookeeper --channel=3/edge
juju relate kafka zookeeper
```

> Deploy the benchmark tool and relate it to the cluster:

Reviewer: **nitpick**: Please put an empty line between each code block and the rest of the content.
```suggestion

Deploy the benchmark tool and relate it to the cluster:

```

```
juju deploy kafka-benchmark --channel=latest/edge
juju relate kafka kafka-benchmark

> ```

Reviewer 2: **todo** can we also add the TLS support? Just a subsection "TLS enabling", I suppose we need to still relate to a tls operator, right?

### Integrate with COS

> The benchmark charm needs COS to collect metrics during the test. Relate the Apache Kafka benchmark with a [grafana-agent operator](https://charmhub.io/grafana-agent).

Reviewer: **Question**: Do we need to point out somewhere, that the COS should run on a different model and ideally, on a different hardware to have less effect on the performance during benchmarking.
I know it seems obvious, especially since COS is on K8s but this detail seems relevant for a benchmark.
Maybe, we can use something like this:
```suggestion
The benchmark charm needs a [COS](https://charmhub.io/topics/canonical-observability-stack) deployment running on a separate Juju model in Kubernetes to collect metrics during the test. Relate the Apache Kafka benchmark with a [grafana-agent operator](https://charmhub.io/grafana-agent).
```

> For more details on how to deploy and configure COS and its agents, check [the upstream documentation](https://canonical.com/data/docs/kafka/iaas/h-enable-monitoring).

Reviewer: **Suggestion**: Let's use more specific text for the link.

```suggestion
For instructions on how to deploy and configure COS and its agents, see the [How to enable monitoring](https://canonical.com/data/docs/kafka/iaas/h-enable-monitoring) guide.
```

Once the grafana-agent is deployed, relate it with:
```
juju relate grafana-agent kafka-benchmark
```
The benchmark data will be collected every 10s and sent to prometheus.

## Run the Benchmark
To kick start a benchmark, execute the following actions:
```

> juju run kafka-benchmark/leader prepare  # to set the environment and the cluster

Reviewer: **Suggestion**: We should avoid in-line comments and better yet any comments in a code block if we can. I'd suggest splitting this code block in to with proper introductory text for each.

juju run kafka-benchmark/leader run
```

The units will pick-up the command and start executing the benchmark.

### Stop the Benchmark
To stop the benchmark, execute:
```
juju run kafka-benchmark/leader stop
```

> Optionally, it is possible to clean the current benchmark data using:

Reviewer: **Questiion**: Just as an idea, can we provide any guidance about the results? 
* Where to get them (I assume in COS)? 
* How to see/retrieve/analyse them with a reference to a Prometheus or COS docs?
* Maybe, even a small screenshot of some chart/diagram that is too small to see any details but provides visualisation of the end goal.

```
juju run kafka-benchmark/leader cleanup
```

> That will return the Apache Kafka benchmark charm to its original condition.

Reviewer: **Question**: Should we add instructions on how to tear down the deployment correctly to free up the resources?
